server/service/apple_mdm.go

```diff
@@ -1334,9 +1346,56 @@ func (svc *Service) GetMDMAppleEnrollmentProfileByToken(ctx context.Context, tok
 		return nil, ctxerr.Wrap(ctx, err)
 	}
 
-	enrollURL, err := apple_mdm.AddEnrollmentRefToFleetURL(appConfig.ServerSettings.ServerURL, ref)
-	if err != nil {
-		return nil, ctxerr.Wrap(ctx, err, "adding reference to fleet URL")
+	enrollURL := appConfig.ServerSettings.ServerURL
+	// check if there is a legacy enroll ref and preserve it for backwards compatibility
+	if ref != "" {
+		idpAcct, err := svc.ds.GetMDMIdPAccountByAccountUUID(
+			// use the primary db as the account might have been just inserted
+			ctxdb.RequirePrimary(ctx, true),
+			ref,
+		)
+		if err != nil {
+			return nil, ctxerr.Wrap(ctx, err, "getting MDM IdP account")
+		}
+		if idpAcct.FleetEnrollRef != "" {
+			level.Debug(svc.logger).Log("msg", "using legacy enroll ref", "enroll_ref", idpAcct.FleetEnrollRef)
+			// we have a legacy enroll ref, add it to the enroll URL
+			enrollURL, err = apple_mdm.AddEnrollmentRefToFleetURL(appConfig.ServerSettings.ServerURL, idpAcct.FleetEnrollRef)
+			if err != nil {
+				return nil, ctxerr.Wrap(ctx, err, "adding reference to fleet URL")
+			}
+		}
+	}
+
+	// TODO: Consider renaming/refactoring this method. We need to do more than just getting the
+	// profile here, we also need to the device with an MDM IdP account
+
+	// if we have device info, we need to parse it and associate the device with an MDM IdP account
+	if deviceinfo != "" {
+		// TODO: Consider implementing just what we need from the parser, instead of needing to
+		// mock up a request that we can pass to the parser
+		r, err := http.NewRequest("", "", nil)
+		if err != nil {
+			return nil, ctxerr.Wrap(ctx, err, "creating request")
+		}
+		level.Info(svc.logger).Log("msg", "deviceinfo", "deviceinfo", deviceinfo)
+		r.Header.Add("x-apple-aspen-deviceinfo", deviceinfo)
+		// extract x-apple-aspen-deviceinfo custom header from request
+		di, err := dep_webview.DefaultParser.Parse(r)
+		if err != nil {
+			return nil, ctxerr.Wrap(ctx, err, "parsing deviceinfo")
+		}
+
+		// try to associate the device with an MDM IdP account
+		if ref == "" {
+			// this is unexpected, there should be a reference if we have device info
+			level.Debug(svc.logger).Log("msg", "associating mdm idp account, missing idp account uuid", "device_uuid", di.UDID)
+		} else {
```

Comment:
I think this answers my question about why we still have the ref lingering around. I was imagining we could completely replace it with the device info. Did you find any blockers?

Reply:
@roperzh, the issue I ran into is that we don't have access to the deviceinfo when the server is handling the SSO callback (because the callback request is coming from the IdP rather than the client at this point) so we still need something to make the linkage.

Reply:
@gillespi314 how is the reference making it through? can we do the same?

Reply:
We create the reference when we insert the MDM IdP account when the server is handling the SSO [callback](https://github.com/fleetdm/fleet/blob/7081c511bd74bba183609c168ca7ef71b0294d4a/ee/server/service/mdm.go#L789-L790). That reference gets tacked on to the redirect URL, which the UI can extract via props. Then the UI pulls from local storage and includes the deviceinfo and the reference when it makes the DEP profile request, which completes the loop. I don't see a good way to swap the reference for the host UUID.

Reply:
The best I can think of would be to create something that links the initial SSO request (where we do have deviceinfo) with the callback request (from the IdP where we don't have deviceinfo). I think that would be the Redis approach rather than local storage.

Reply:
understood, then it sounds good to me! if we can't find an easy way to get rid of it, just removing it from the enrollment profile is a major win